File: android/app/src/main/java/com/mapper/env/AppEnvModule.kt
package com.mapper.env

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.module.annotations.ReactModule
import com.mapper.BuildConfig
import com.mapper.R

/**
 * Đưa cấu hình theo flavor từ BuildConfig sang JS.
 *
 * Vì sao không dùng react-native-config: lib đó phải đọc .env lúc build, thêm
 * một nguồn sự thật nữa, và là lib có rủi ro New Architecture cao nhất trong
 * danh sách ở docs/00 mục 3. BuildConfig đã có sẵn giá trị theo flavor, chỉ cần
 * một module ~30 dòng để đọc ra — không thêm dependency nào.
 *
 * Giá trị trả về là CONSTANT (getConstants) nên JS đọc đồng bộ ngay lúc import,
 * không cần await, không có race lúc khởi động.
 */
@ReactModule(name = AppEnvModule.NAME)
class AppEnvModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

  companion object {
    const val NAME = "MapperAppEnv"
  }

  override fun getName(): String = NAME

  override fun getConstants(): Map<String, Any> {
    val ctx = reactApplicationContext
    return mapOf(
        "flavor" to BuildConfig.FLAVOR_ENV,
        "apiBaseUrl" to BuildConfig.API_BASE_URL,
        "forgeRockUrl" to BuildConfig.FORGEROCK_URL,
        "forgeRockRealm" to BuildConfig.FORGEROCK_REALM,
        "sePayEnv" to BuildConfig.SEPAY_ENV,
        "widgetRefreshMinutes" to BuildConfig.WIDGET_REFRESH_MINUTES,
        "applicationId" to BuildConfig.APPLICATION_ID,
        "versionName" to BuildConfig.VERSION_NAME,
        "buildNumber" to BuildConfig.VERSION_CODE,
        "isDebug" to BuildConfig.DEBUG,
        "appName" to ctx.getString(R.string.app_name),
    )
  }
}

File: ios/Mapper/Native/AppEnvModule.swift
import Foundation

/// Đưa cấu hình theo flavor từ Info.plist (đến từ xcconfig) sang JS.
///
/// Đối xứng với `AppEnvModule.kt` bên Android: cùng tên module `MapperAppEnv`,
/// cùng bộ key, nên `src/config/env.ts` không cần `Platform.select`.
///
/// Dùng `constantsToExport` để JS đọc đồng bộ ngay lúc import — không có race
/// lúc khởi động, không cần await.
@objc(MapperAppEnv)
class MapperAppEnv: NSObject {

  private func str(_ key: String, _ fallback: String = "") -> String {
    (Bundle.main.object(forInfoDictionaryKey: key) as? String) ?? fallback
  }

  @objc
  func constantsToExport() -> [AnyHashable: Any]! {
    let host = str("ApiHost")
    let frHost = str("ForgeRockHost")
    let refreshMinutes = Int(str("WidgetRefreshMinutes", "5")) ?? 5

    return [
      "flavor": str("FlavorEnv", "prod"),
      // Cổng 443 là mặc định của https — ghi ":443" vào URL chỉ làm rối log và
      // dễ gây lệch khi so origin. Giữ nguyên dạng chuẩn.
      "apiBaseUrl": "https://\(host)",
      "forgeRockUrl": "https://\(frHost)/am",
      "forgeRockRealm": str("ForgeRockRealm", "alpha"),
      "sePayEnv": str("SePayEnv", "production"),
      "widgetRefreshMinutes": refreshMinutes,
      "applicationId": Bundle.main.bundleIdentifier ?? "",
      "versionName": str("CFBundleShortVersionString", "0.0.0"),
      "buildNumber": Int(str("CFBundleVersion", "0")) ?? 0,
      "isDebug": MapperAppEnv.isDebugBuild,
      "appName": str("CFBundleDisplayName", "Mapper"),
      "appGroupId": str("AppGroupId"),
      "appScheme": str("AppScheme"),
    ]
  }

  private static var isDebugBuild: Bool {
    #if DEBUG
      return true
    #else
      return false
    #endif
  }

  @objc
  static func requiresMainQueueSetup() -> Bool { false }
}
